### ISSUE TITLE
Unable to reproduce the results

### ISSUE THREAD

**Issue by puneater** [GMvandeVen/brain-inspired-replay#7]
Hello,

Sorry to bug you, but I think I am unable to reproduce the results.

Check out my notebook,
https://colab.research.google.com/drive/1g55JOiJnFRsTB7kjqkJO9YykTy1JDHhm?usp=sharing

I am trying to reproduce results on Class-IL for EWC and LWF for 10 tasks. I saw in your article that for 10 classes LWF and EWC get higher accuracy than I am able to achieve. Am I doing something wrong?

Great work though!

**Comment by GMvandeVen** [comment]
Thanks!

I had a quick look at your notebook, and one difference with my experiments is that I used pre-trained convolutional layers for the experiments on CIFAR-100. I think that probably explains the difference in results.
You can use the `--pre-convE` flag to use pre-trained convolutional layers. By default it should then use the same pre-trained layers as I used in my study.

Hope this helps!

**Comment by puneater** [comment]
Thank you for the quick response, I'll definitely try that and get back to you.

**Comment by puneater** [comment]
@GMvandeVen Hey, can you please check these screenshots, I tried adding the flag, but I think it's not working.
What am I doing wrong here?

![image](https://user-images.githubusercontent.com/45456578/106751073-bc71f800-664e-11eb-95d8-a2f1866a8a41.png)


![image](https://user-images.githubusercontent.com/45456578/106751127-ce539b00-664e-11eb-9564-69ad96bc1d9f.png)

**Comment by GMvandeVen** [comment]
Hi, I had a quick look at your screenshots, but the results you are getting seem reasonable to me. I also found that EWC will perform poorly in the class-incremental learning scenario.
But I did notice another difference with the code you're running and the experiments I reported in the paper, and that is that you are using `--task=5` for CIFAR-100, while I used `--task=10`. That's perhaps it?

**Comment by puneater** [comment]
Hello @GMvandeVen ,

In the article, [figure 6c (bottom right)](https://www.nature.com/articles/s41467-020-17866-2/figures/6) depicts ≥0.4 test accuracy (approximately) of EWC and LWF on atleast first 0-20 classes they learnt of CIFAR100 and that is what I am unable to reproduce. Am I missing something in the article or reading it wrong?

I will try increasing the tasks too.

Thank you for replying.

**Comment by GMvandeVen** [comment]
Thanks for the clarification, now I understand the confusion.

The test accuracies that are plotted in the figures in the paper are different from the test accuracies that are printed to the screen by my code. The accuracies printed to the screen are for the final model (i.e., after it is trained on all tasks), but the accuracies in the figures in the paper are for the model after it is trained up to a certain task (and with the average accuracy only based on tasks seen so far). This second set of accuracies are not printed out to the screen by my code, but they can be plotted if you use the flag `--pdf` when calling `./main_cl.py` (second plot in the resulting pdf).

Sorry for the confusion. I also just updated the information that is printed to the screen to better reflect these subtleties.

**Comment by puneater** [comment]
Oh, now I get it.

One more doubt, probably the last one, the accuracy that is printed in the paper upto a certain task, is that the accuracy on the n<sup>th</sup> task the neural network just learnt or is it the accuracy on all n tasks? (I'm guessing it's the former one)

Thank you for all the help @GMvandeVen , you're great!

I will use that flag, and check with the results.

Thanks again.

**Comment by GMvandeVen** [comment]
Great, thanks! Of the two options you give, it's actually the latter. The accuracy after _n_ tasks that is plotted in the figures in the paper is the average accuracy over all _n_ tasks so far.

**Comment by puneater** [comment]
Oh okay, so in that figure 6c, the initial accuracy of ~0.8, is that on the classes the neural network pretrained on or are those 10 classes also learnt one by one using EWC?

![image](https://user-images.githubusercontent.com/45456578/107371019-c0e45800-6b09-11eb-9591-0403d0eca10e.png)

**Comment by GMvandeVen** [comment]
On the first task (i.e., the first 10 classes), there is no EWC regularization yet. So on this task you could say that the network is trained "in the standard way".
What in the paper is referred to as "pre-training" is before this first task. Pre-training was on CIFAR-10, but no accuracy on any CIFAR-10 classes is reported.

**Comment by puneater** [comment]
Yes, I understand now.
So if I execute the following,
`python main_cl.py --ewc --lambda=5000 --experiment CIFAR100 --scenario class --tasks 5 --pre-convE`
Will it divide the classes in 5 tasks, that is, 20 classes per task or will it take 10 classes per task or is it 1 class per task?

**Comment by GMvandeVen** [comment]
With that command, CIFAR-100 will be divided into 5 tasks with 20 classes each. The line that sets the number of classes per task is here: https://github.com/GMvandeVen/brain-inspired-replay/blob/ce30785bd1451b768bacb17c819a4236cd06389a/data/load.py#L159

**Comment by puneater** [comment]
Thank you very much, @GMvandeVen , for all the help.
I understand the experiments now.

And sorry for bugging you this much.
I'm closing this thread.

Thanks again!
